torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -1179,6 +1175,36 @@ def call_triton_kernel(
                 "Please use a Config in @triton.autotune instead."
             )
 
+        # Run prune_configs_by to filter the configs
+        if isinstance(variable.kernel, Autotuner):
+            # args and kwargs aren't real values yet
+            # e.g., LazyVariableTracker
+            # If a user wants to access them, we should realize them
+            def realize_var(
+                var: VariableTracker,
+            ) -> Union[Any, VariableTracker, LazyVariableTracker]:
```

Comment:
What's the high level strategy here? Are we  "realizing" the configs, then running the prune_configs_by function, then storing that information somewhere?

Reply:
@zou3519 in this PR, we want to add support for the `prune_configs_by` parameter of the `@triton.autotune` decorator (see the docs [here](https://triton-lang.org/main/python-api/generated/triton.autotune.html#triton.autotune)). (@SamGinzburg could you please update the PR head with more context?)

The implementation relies on the `Autotuner.prune_configs` method ([here](https://github.com/triton-lang/triton/blob/73ba8b639226fbc505115db59189a1362c350075/python/triton/runtime/autotuner.py#L214)) which runs the user-provided callbacks (`early_config_prune` and `perf_model`) on the `args` and `kwargs` passed to the kernel and prunes the configs in-place (i.e., resets the `Autotuner.configs` attribute to a smaller list). So two important things here are:

- We should set `args` as `Autotuner.nargs` and pass `kwargs` *separately*.

- The original `args` and `kwargs` values (not variable trackers) should be passed to the callbacks, so we should realize before passing.

We've added the logic in the triton_kernel_wrap, because the original `args` and `kwargs` are still separated here. To make the code generic wrt. Dynamo vs non-strict export, we could add a helper method `TritonHOPifier.realize` for realizing the values, which would be implemented differently in the `DynamoTritonHOPifier` and `TracingTritonHOPifier` subclasses. Then we can call `self.realize` in the `TritonHOPifier.call_triton_kernel` and dispatch to the implementation-specific version of realization. Does this make sense to you?

But I also have two questions re. specific implementations of `realize`:

1. Is the Dynamo-specific realization code above valid or are we missing smth / is there a better way?

2. What does realization look like for non-strict export? Is it a no-op (i.e., the values are the same as passed to the kernel call) or something needs to be done for realizaiton?

Thanks!

Reply:
> Is the Dynamo-specific realization code above valid or are we missing smth / is there a better way?

We usually don't unpack values back into their eager types inside of Dynamo. Also, this is only really OK to do if all of the types are constants (that is, Dynamo guards on their values), and if the `early_config_prune`/`perf_model`/other user-defined functions are "safe to execute" (that is, it has no side effects, which is really not guaranteed here).

The alternative strategy that you probably want to do instead is to create a function in Dynamo and then invoke it with the VariableTracker arguments, so that Dynamo executes the function via its bytecode interpreter (instead of cpython's). Something like the following:
```py
def do_prune_configs(early_config_prune, perf_model, configs_top_k, nargs, configs, kwargs):
    autotuner = object()
    autotuner.early_config_prune = early_config_prune
    autotuner.perf_model = perf_model
    autotuner.configs = configs
    autotuner.configs_top_k = configs_top_k
    configs = AutoTuner.prune_configs(autotuner, kwargs)
    return configs

UserDefinedFunctionVariable(do_prune_configs).call_function(tx, variable_trackers)
```

Reply:
I'm a bit worried about duck typing of the `Autotuner` in the above: if the implementation details in the Triton code change, this may break. What would be the `variable_trackers` passed to the `UserDefinedFunctionVariable(do_prune_configs).call_function`? Because we have `early_config_prune` and `perf_model` as Python functions on the `Autotuner` object, not variable trackers. If passing normal Python objects to `UserDefinedFunctionVariable(do_prune_configs).call_function` is allowed, can we just pass our `autotuner` object instead and call the `autotuner.prune_cofigs(kwargs)` directly? The call will modify the state of the `autotuner` object, but it is not tracked by Dynamo.

Reply:
Alternatively, we can try implementing `prune_configs_by` support in Inductor, as we'll have access to the `Autotuner` object there. We'll need to pass the `len(args)` which we can do through patching the `Autotuner` object in Dynamo and de-patching in Inductor.